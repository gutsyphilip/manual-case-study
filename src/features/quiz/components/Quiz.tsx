import React, { Component } from "react";
import { Modal } from "../../../components/modal";
import Question from "./Question";
import {
  ManualLogo,
  FrownIcon,
  SmileIcon,
  CloseIcon,
} from "../../../assets/svgs";

import { questions } from "../../../data/quiz.json";

import styles from "./quiz.module.scss";

interface Props {
  setShowQuizModal: (value: boolean) => void;
}
interface State {
  activeQuestion: number;
  userResponses: { [key: string]: string | boolean };
  isCompleted: boolean;
}

class Quiz extends Component<Props, State> {
  state = {
    activeQuestion: 1,
    userResponses: {},
    isCompleted: false,
  };

  handleSetResponse = (
    question: string,
    response: string | boolean,
    questionNumber: number
  ) => {
    const { userResponses } = this.state;
    this.setState({
      userResponses: { ...userResponses, [question]: response },
    });
    if (questionNumber < 3) {
      this.setState({ activeQuestion: questionNumber + 1 });
    } else {
      this.handleSubmitQuiz();
    }
  };

  handleSubmitQuiz = () => {
    this.setState({ isCompleted: true });
  };

  render() {
    const { activeQuestion, isCompleted, userResponses } = this.state;
    const { setShowQuizModal } = this.props;
    const activeRef = React.createRef();
    return (
      <Modal
        className={styles.quizModal}
        size="cover"
        show={true}
        {...this.props}
      >
        <header className={styles.quizModal_header}>
          <ManualLogo className={styles.quizModal_header_logo} />
          <CloseIcon
            className={styles.quizModal_header_closeIcon}
            onClick={() => {
              setShowQuizModal(false);
            }}
          />
        </header>
        {!isCompleted &&
          questions.map((item, index) => {
            return (
              <Question
                userResponses={userResponses}
                activeRef={activeRef}
                handleSetResponse={this.handleSetResponse}
                key={item.question.replace(" ", "+")}
                data={item}
                number={index + 1}
                activeQuestion={activeQuestion}
              />
            );
          })}

        {isCompleted && (
          <section className={styles.quizModal_result}>
            {Object.values(userResponses).includes(true) ? (
              <>
                <FrownIcon className={styles.quizModal_result_icon} />
                <p className={styles.quizModal_result_message}>
                  Unfortunately, we are unable to prescribe this medication for
                  you. This is because finasteride can alter the PSA levels,
                  which maybe used to monitor for cancer. You should discuss
                  this further with your GP or specialist if you would still
                  like this medication.
                </p>
              </>
            ) : (
              <>
                <SmileIcon className={styles.quizModal_result_icon} />
                <p className={styles.quizModal_result_message}>
                  Great news! We have the perfect treatment for your hair loss.
                  Proceed to{" "}
                  <a
                    href="https://www.manual.co/"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    www.manual.co
                  </a>
                  , and prepare to say hello to your new hair!
                </p>
              </>
            )}
          </section>
        )}
      </Modal>
    );
  }
}

export default Quiz;
