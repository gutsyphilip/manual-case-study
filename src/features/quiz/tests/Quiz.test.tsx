import React, { useRef } from "react";
import { render, cleanup } from "@testing-library/react";
import Quiz from "../components/Quiz";
import { Simulate } from "react-dom/test-utils";

const QuizProps = {
  setShowQuizModal: jest.fn(() => {}),
  showQuizModal: true,
};

test("quiz modal renders properly", () => {
  const modalRoot = document.createElement("div");
  modalRoot.setAttribute("id", "modal");
  const { container, unmount } = render(<Quiz {...QuizProps} />, {
    container: document.body.appendChild(modalRoot),
  });
  expect(container).toMatchSnapshot();
  unmount();
});

test("user can select quiz options", () => {
  const modalRoot = document.createElement("div");
  modalRoot.setAttribute("id", "modal");
  const { container, getAllByTestId, unmount } = render(
    <Quiz {...QuizProps} />,
    {
      container: document.body.appendChild(modalRoot),
    }
  );
  const optionsListNodes = getAllByTestId("options-list");
  //   console.log(optionsListNodes[0].firstChild);
  optionsListNodes.forEach((options) => {
    // console.log(options.firstChild);
    Simulate.click(options.firstChild);
  });
  console.log(container);
  unmount();

  //   expect(questionNode).toMatchSnapshot();
});
