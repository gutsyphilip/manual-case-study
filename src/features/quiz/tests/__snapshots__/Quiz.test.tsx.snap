// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quiz modal renders properly 1`] = `
<div
  id="modal"
>
  <div>
    <div
      aria-label="Quiz modal"
      class="modal quizModal modal_cover"
      role="dialog"
    >
      <main
        class="modal_body"
      >
        <header
          class="quizModal_header"
        >
          <svg
            class="quizModal_header_logo"
          >
            manual-logo.svg
          </svg>
          <svg
            class="quizModal_header_closeIcon"
          >
            close.svg
          </svg>
        </header>
        <div
          class="quizModal_question"
        >
          <h4
            class="quizModal_question_label"
          >
            Question 1
          </h4>
          <h1
            class="quizModal_question_title"
          >
            Which image best matches your hair loss?
          </h1>
          <div
            class="quizModal_question_options"
            data-testid="options-list"
          >
            <div
              class="quizModal_question_options_item"
              data-testid="option"
            >
              <div
                class="quizModal_question_options_item_label"
              >
                <img
                  alt="Temples"
                  src="https://s3-eu-west-1.amazonaws.com/manualco/questions/temples-hairloss.png"
                  srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/temples-hairloss%402x.png 2x"
                />
              </div>
              <div
                class="quizModal_question_options_item_text"
              >
                Temples
              </div>
            </div>
            <div
              class="quizModal_question_options_item"
              data-testid="option"
            >
              <div
                class="quizModal_question_options_item_label"
              >
                <img
                  alt="Temples & Crown"
                  src="https://s3-eu-west-1.amazonaws.com/manualco/questions/templescrown-hairloss.png"
                  srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/templescrown-hairloss%402x.png 2x"
                />
              </div>
              <div
                class="quizModal_question_options_item_text"
              >
                Temples & Crown
              </div>
            </div>
            <div
              class="quizModal_question_options_item"
              data-testid="option"
            >
              <div
                class="quizModal_question_options_item_label"
              >
                <img
                  alt="Patchy"
                  src="https://s3-eu-west-1.amazonaws.com/manualco/questions/patchy-hairloss.png"
                  srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/patchy-hairloss%402x.png 2x"
                />
              </div>
              <div
                class="quizModal_question_options_item_text"
              >
                Patchy
              </div>
            </div>
            <div
              class="quizModal_question_options_item"
              data-testid="option"
            >
              <div
                class="quizModal_question_options_item_label"
              >
                <img
                  alt="Moderate"
                  src="https://s3-eu-west-1.amazonaws.com/manualco/questions/moderate-hairloss.png"
                  srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/moderate-hairloss%402x.png 2x"
                />
              </div>
              <div
                class="quizModal_question_options_item_text"
              >
                Moderate
              </div>
            </div>
            <div
              class="quizModal_question_options_item"
              data-testid="option"
            >
              <div
                class="quizModal_question_options_item_label"
              >
                <img
                  alt="Extensive"
                  src="https://s3-eu-west-1.amazonaws.com/manualco/questions/extensive-hairloss.png"
                  srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/extensive-hairloss%402x.png 2x"
                />
              </div>
              <div
                class="quizModal_question_options_item_text"
              >
                Extensive
              </div>
            </div>
            <div
              class="quizModal_question_options_item"
              data-testid="option"
            >
              <div
                class="quizModal_question_options_item_label"
              >
                <img
                  alt="Complete"
                  src="https://s3-eu-west-1.amazonaws.com/manualco/questions/complete-hairloss.png"
                  srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/complete-hairloss%402x.png 2x"
                />
              </div>
              <div
                class="quizModal_question_options_item_text"
              >
                Complete
              </div>
            </div>
          </div>
        </div>
        <div
          class="quizModal_question"
        >
          <h4
            class="quizModal_question_label"
          >
            Question 2
          </h4>
          <h1
            class="quizModal_question_title"
          >
            Have you ever been diagnosed with prostate cancer, or are you currently undergoing PSA/Prostate monitoring?
          </h1>
          <div
            class="quizModal_question_options"
            data-testid="options-list"
          >
            <div
              class="quizModal_question_options_item"
              data-testid="option"
            >
              <div
                class="quizModal_question_options_item_label"
              >
                Yes
              </div>
              <div
                class="quizModal_question_options_item_text"
              />
            </div>
            <div
              class="quizModal_question_options_item"
              data-testid="option"
            >
              <div
                class="quizModal_question_options_item_label"
              >
                No
              </div>
              <div
                class="quizModal_question_options_item_text"
              />
            </div>
          </div>
        </div>
        <div
          class="quizModal_question"
        >
          <h4
            class="quizModal_question_label"
          >
            Question 3
          </h4>
          <h1
            class="quizModal_question_title"
          >
            Have you ever been diagnosed with breast cancer or noticed any changes in  your breast tissue such as lumps, pain, nipple discharge or swelling?
          </h1>
          <div
            class="quizModal_question_options"
            data-testid="options-list"
          >
            <div
              class="quizModal_question_options_item"
              data-testid="option"
            >
              <div
                class="quizModal_question_options_item_label"
              >
                Yes
              </div>
              <div
                class="quizModal_question_options_item_text"
              />
            </div>
            <div
              class="quizModal_question_options_item"
              data-testid="option"
            >
              <div
                class="quizModal_question_options_item_label"
              >
                No
              </div>
              <div
                class="quizModal_question_options_item_text"
              />
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
